src/main/ruby/truffleruby/core/regexp.rb

```diff
@@ -144,8 +144,30 @@ def self.new(pattern, opts = undefined, encoding = nil)
 
     if Primitive.is_a?(opts, Integer)
       opts = opts & (OPTION_MASK | KCODE_MASK) if opts > 0
-    elsif !Primitive.undefined?(opts) && opts
-      opts = IGNORECASE
+    elsif Primitive.is_a?(opts, String)
+      opts = opts.chars.reduce(0) do |result, opt|
+        case opt
+        when 'i'
+          result | IGNORECASE
+        when 'm'
+          result | MULTILINE
+        when 'x'
+          result | EXTENDED
+        else
+          raise ArgumentError, "unknown regexp option: #{opt}"
+        end
+      end
+    elsif !Primitive.undefined?(opts)
+      # Valid values are true, false, nil.
+      # Other values warn but treat as true.
+      if Primitive.false?(opts) || Primitive.nil?(opts)
+        opts = 0
+      else
+        if !Primitive.true?(opts)
+          warn "expected true or false as ignorecase: #{opts}"
+        end
```

Comment:
This seems like a good place to use suffix `unless`: `warn unless Primitive.true?(opts)`